source/grid/grid_out.cc

```diff
@@ -2590,51 +2590,70 @@ void GridOut::write_mesh_per_processor_as_vtu (const Triangulation<dim,spacedim>
       patches.push_back(patch);
     }
 
-  const std::string new_file = (filename_without_extension + ".proc" +
-                                Utilities::int_to_string (tria.locally_owned_subdomain(), 4) +
-                                ".vtu");
-  std::ofstream out(new_file.c_str());
-  std::vector<std::tuple<unsigned int, unsigned int, std::string> > vector_data_ranges;
-  DataOutBase::VtkFlags flags;
-  DataOutBase::write_vtu (patches,
-                          data_names,
-                          vector_data_ranges,
-                          flags,
-                          out);
-  //create .pvtu record
-  if (tria.locally_owned_subdomain() == 0)
+  // only create .pvtu file if running in parallel
+  // if not, just create a .vtu file with no reference
+  // to the processor number
+  if (const parallel::Triangulation<dim,spacedim> *tr =
+        dynamic_cast<const parallel::Triangulation<dim,spacedim> *>(&tria))
+    {
+      const std::string new_file = (filename_without_extension + ".proc" +
+                                    Utilities::int_to_string (tr->locally_owned_subdomain(), 4) +
+                                    ".vtu");
+      std::ofstream out(new_file.c_str());
+      std::vector<std::tuple<unsigned int, unsigned int, std::string> > vector_data_ranges;
+      DataOutBase::VtkFlags flags;
+      DataOutBase::write_vtu (patches,
+                              data_names,
+                              vector_data_ranges,
+                              flags,
+                              out);
+      // create .pvtu record
+      if (tr->locally_owned_subdomain() == 0)
+        {
+          std::vector<std::string> filenames;
+
+          // .pvtu needs to reference the files without a relative path because it will be written
+          // in the same directory. For this, remove any paths from filename.
+          std::size_t pos = filename_without_extension.find_last_of('/');
+          if (pos == std::string::npos)
+            pos = 0;
+          else
+            pos += 1;
+          for (unsigned int i=0; i<Utilities::MPI::n_mpi_processes(tr->get_communicator()); ++i)
```

Comment:
This is a nice fix burried in this patch :)

Reply:
Please also save `Utilities::MPI::n_mpi_processes(tr->get_communicator())` to a `const` variable to avoid the compiler calculating the number of processes each iteration anew.